### PR TITLE
[FEATURE] Implement Multi-Column Domain Builder for Rule-Based Profiler

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Implement `MultiColumnDomainBuilder`
* Add a unit test for `MultiColumnDomainBuilder`
* Use `safe deep copy` when handling `batch_request` for Rule-Based Profiler (to avoid pickling errors for dynamic data)
* Clean-up existing unit tests

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-464/GREAT-720
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Review by anthonyburdi — APPROVED** [review]
LGTM!

**Review by Shinnnyshinshin — APPROVED** [review]
Love it. Can we add a to do to add a use-case to the Notebook, either in this PR or the next?

File: great_expectations/rule_based_profiler/domain_builder/multi_column_domain_builder.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            include_semantic_types=None,
+            exclude_semantic_types=None,
+            batch_list=batch_list,
+            batch_request=batch_request,
+            data_context=data_context,
+        )
+
+    @property
+    def domain_type(self) -> Union[str, MetricDomainTypes]:
+        return MetricDomainTypes.MULTICOLUMN
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/domain_builder/multi_column_domain_builder.py:46]: I know we accept string representations but is this not a single return type of `MetricDomainTypes`?

File: tests/rule_based_profiler/domain_builder/test_domain_builder.py [hunk 2]
```diff
@@ -168,19 +164,85 @@ def test_column_domain_builder_with_simple_semantic_type_included(
...
+                "inferred_semantic_domain_type": "numeric",
+            },
         },
     ]
+
+
+def test_multi_column_domain_builder(
+    alice_columnar_table_single_batch_context,
+    alice_columnar_table_single_batch,
+):
```
Comment (cdkini) [inline tests/rule_based_profiler/domain_builder/test_domain_builder.py:193]: I think Alice/Bobby/Bobster/Quentin should be moved to `integration` and we should write standard unit tests here.

**Review by cdkini — APPROVED** [review]
LGTM - minor comments but not blocking

File: great_expectations/rule_based_profiler/domain_builder/multi_column_domain_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/multi_column_domain_builder.py:46]: @cdkini Just adhering the the interface.  If you think we should change the interface, let us discuss -- if no implications, fine with me (but not sure right now).  Thanks!

**Pull request merged**